File: projects/engineering.purdue.edu/scriptingwithobjects/swocode/chap17/XmlRpcDaemonServer.pl
#!/usr/bin/perl

### XmlRpcDaemonServer.pl

use Frontier::Daemon;                                                #(A)

sub sumAndDifference {                                               #(B)
    my ($x, $y) = @_;                                                #(C)
#    my $arrayref = shift;
#    my ($x, $y) = @$arrayref;
#    print "server x: $x   server y: $y\n";
    return {'sum' => $x + $y, 'difference' => $x - $y};              #(D)
}

sub sayHello {                                                       #(E)
    my $sender = shift;
    return "Hello to $sender from an XML-RPC daemon server\n";
}

$methods = {'sumAndDifference' => \&sumAndDifference,
            'sayHello'         => \&sayHello,
           };                                                        #(F)

# From local host, call me at http://localhost:8080/RPC2
# From a remote host, call me at http://128.46.144.238:8080/RPC2
Frontier::Daemon->new( LocalPort => 8080, 
                       methods => $methods,
                     )                                        
    or die "Couldn't start HTTP server: $!";                         #(G)

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap17/XmlRpcDaemonServer.py
#!/usr/bin/env python

### XmlRpcDaemonServer.py

import SimpleXMLRPCServer                                            #(A)

def sumAndDifference( arg1, arg2 ):                                  #(B)
    return { 'sum' : arg1 + arg2, 'difference' : arg1 - arg2 }       #(C)

def sayHello( sender ):                                              #(D)
    return "Hello to %s from an XML-RPC daemon server" % sender      #(E)

server = SimpleXMLRPCServer.SimpleXMLRPCServer(('localhost', 8080))  #(F)

server.register_function( sumAndDifference )                         #(G)
server.register_function( sayHello )                                 #(H)
server.register_introspection_functions()                            #(I)

server.serve_forever()                                               #(J)
